### PR TITLE
feat(get-platform): remove error, rephrase it as a warning

### LINKED ISSUE #17390: Regression: `get-platform` error for non-`amd64` Alpine introduced in `prisma@4.9.0` prevents using custom Prisma engines
As stated by [this comment](https://github.com/prisma/prisma/pull/17334#issuecomment-1386451802), the new error introduced for non-`amd64` Linux Alpine in https://github.com/prisma/prisma/pull/17334 prevents users from using their own custom engines, (which e.g. may be compiled explicitly for `arm64` to circumvent https://github.com/prisma/prisma/issues/8478).

From [Slack](https://prisma-company.slack.com/archives/C04DG9JL6TV/p1674029182666039):
> That isn’t something we’ve ever supported explicitly or told people to do, is it?

> I think it is, we have a page in documentation about using custom engines (although not about building them), and we also told people in that issue that building the engines themselves is one of the workarounds

### PR DESCRIPTION
**Pull request by jkomyno**
Closes https://github.com/prisma/prisma/issues/17390.

- [x] Stop erroring out when using an unsupported arch when running Prisma on Linux Alpine
- [x] Add link CLI utility to `@prisma/get-platform`

## TODO

- [ ] Create a pris.ly link `https://pris.ly/d/docker-alpine` to the tutorial generated by https://github.com/prisma/docs/issues/4365

### CODE REVIEW

File: packages/get-platform/src/getPlatform.ts [hunk 1]
```diff
@@ -55,14 +55,27 @@ export async function getos(): Promise<GetOSResult> {
   }
 
   const distro = await resolveDistro()
+  const archFromUname = await getArchFromUname()
 
+  const usesOfficialEngines = !(
+    process.env.PRISMA_QUERY_ENGINE_BINARY ||
+    process.env.PRISMA_QUERY_ENGINE_LIBRARY ||
+    process.env.PRISMA_MIGRATION_ENGINE_BINARY
+  )
```
Comment (jkomyno) [inline packages/get-platform/src/getPlatform.ts:64]: Notes:
- `PRISMA_INTROSPECTION_ENGINE_BINARY` is now ignored
- We need to define the env vars as literals as we can't import `resolveBinary.ts` utils defined in `@prisma/internals` (it would add circular dependencies)

File: packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts [hunk 2]
```diff
@@ -72,6 +72,11 @@ describe('parseLibSSLVersion', () => {
       content: `/lib/libssl.so.3.1`,
       expect: '3.0.x',
     },
+    {
+      name: 'undefined',
+      content: `/lib/libssl3.so`,
+      expect: undefined,
+    },
```
Comment (jkomyno) [inline packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:79]: Random missing test case I had in the drafts locally

**Comment by jkomyno** [comment]
Integration PR: https://github.com/prisma/prisma/pull/17392

File: packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts (hunk 2 above)
Comment (aqrln) [reply packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:79]: maybe also add a comment

```ts
// nss library must not be mistakenly parsed as openssl
```

Comment (jkomyno) [inline packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:79]:
```suggestion
    {
      name: 'The "nss" library must not be mistakenly parsed as "libssl"',
      content: `/lib/libssl3.so`,
      expect: undefined,
    },
```

Comment (jkomyno) [reply packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:79]: I'll add it as the `name` value directly

**Comment by jkomyno** [comment]
Decision points:

1. Is this enough to detect whether Prisma is being run via custom compiled engines?

  ```typescript
    const usesOfficialEngines = !(
      process.env.PRISMA_QUERY_ENGINE_BINARY ||
      process.env.PRISMA_QUERY_ENGINE_LIBRARY ||
      process.env.PRISMA_MIGRATION_ENGINE_BINARY
    )
  ```

2. It's easier to just output a warning and avoid the error altogether, as users realistically want to use `alpine` on `amd64` (already supported) and `arm64` (will be supported by `prisma@4.10.0`)

File: packages/get-platform/src/link.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import chalk from 'chalk'
+import terminalLink from 'terminal-link'
+
+export function link(url): string {
+  return terminalLink(url, url, {
+    fallback: (url) => chalk.underline(url),
+  })
+}
```
Comment (jkomyno) [inline packages/get-platform/src/link.ts:8]: This is currently copied from `@prisma/internals` due to circular dependencies concerns.

File: packages/get-platform/src/getPlatform.ts [hunk 4]
```diff
@@ -55,14 +56,18 @@ export async function getos(): Promise<GetOSResult> {
   }
 
   const distro = await resolveDistro()
+  const archFromUname = await getArchFromUname()
 
-  if (distro === 'musl' && arch !== 'x64') {
-    throw new Error(
-      `Prisma only supports Linux Alpine on the amd64 (x86_64) system architecture. If you're running Prisma on Docker, please use Docker Buildx to simulate the amd64 architecture on your device as explained by this comment: https://github.com/prisma/prisma/issues/8478#issuecomment-1355209706`,
+  // TODO: add 'arm64' to the `[...].includes(arch)` check once we have arm64 engines for Alpine
+  if (distro === 'musl' && !['amd64'].includes(arch)) {
```
Comment (millsp) [inline packages/get-platform/src/getPlatform.ts:62]:
```suggestion
  if (distro === 'musl' && !['amd64', 'x64'].includes(arch)) {
```

File: packages/get-platform/src/getPlatform.ts [hunk 5]
```diff
@@ -55,14 +56,18 @@ export async function getos(): Promise<GetOSResult> {
...
+    warnOnce(
+      'alpine:unsupported-arch',
+      `Prisma only officially supports Linux Alpine on the amd64 (x86_64) system architecture. If you are using your own custom Prisma engines, you can ignore this warning, as long as you've compiled the engines for your system architecture "${archFromUname}".
```
Comment (millsp) [inline packages/get-platform/src/getPlatform.ts:65]: Is Alpine the only distro using `musl`?

File: packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts [hunk 6]
```diff
@@ -72,6 +72,11 @@ describe('parseLibSSLVersion', () => {
       content: `/lib/libssl.so.3.1`,
       expect: '3.0.x',
     },
+    {
```
Comment (millsp) [inline packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:75]: Looks unrelated?

Comment (jkomyno) [reply packages/get-platform/src/__tests__/parseOpenSSLVersion.test.ts:75]: It was a random missing test: https://github.com/prisma/prisma/pull/17389#discussion_r1073363280

File: packages/get-platform/src/getPlatform.ts (hunk 4 above)
Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts:62]: Fixed

Comment (millsp) [reply packages/get-platform/src/getPlatform.ts]: Quick question: Would this have failed with a test somewhere?

Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts]: Currently, no. https://github.com/prisma/ecosystem-tests/issues/3332 would add assertions on some messages in `ecosystem-tests / docker`, we could catch this sort of things by performing assertions on warnings as well.

**Review by Jolg42 — CHANGES_REQUESTED** [review]
@jkomyno Could you remove this binary `packages/engines/introspection-engine-darwin-arm64` and force push?

**Review by janpio — APPROVED** [review]

File: packages/get-platform/src/getPlatform.ts [hunk 7]
```diff
@@ -71,14 +72,18 @@ export async function getos(): Promise<GetOSResult> {
...
 
-  if (distroInfo.targetDistro === 'musl' && arch !== 'x64') {
-    throw new Error(
-      `Prisma only supports Linux Alpine on the amd64 (x86_64) system architecture. If you're running Prisma on Docker, please use Docker Buildx to simulate the amd64 architecture on your device as explained by this comment: https://github.com/prisma/prisma/issues/8478#issuecomment-1355209706`,
+  // TODO: add 'arm64' to the `[...].includes(arch)` check once we have arm64 engines for Alpine
+  if (distroInfo.targetDistro === 'musl' && !['x64'].includes(arch)) {
+    warnOnce(
+      'alpine:unsupported-arch',
+      `Prisma only officially supports Linux Alpine on the amd64 (x86_64) system architecture. If you are using your own custom Prisma engines, you can ignore this warning, as long as you've compiled the engines for your system architecture "${archFromUname}".
+If you are using Prisma on Docker, please refer to ${link('https://pris.ly/d/docker-alpine')}`,
```
[the remark this answers is not in the document]
Comment (jkomyno) [reply packages/get-platform/src/getPlatform.ts:82]: Indeed, it's mentioned in the `TODO` at the top of this PR and in the `Docker` issue in docs

**Review by Jolg42 — APPROVED** [review]

**Review by millsp — APPROVED** [review]
